### ISSUE TITLE
`ReadManyItemsAsync` causes threadpool starvation on low-end App Service (B1)

### ISSUE THREAD

**Issue by tiesmaster** [Azure/azure-cosmos-dotnet-v3#3665]
**Describe the bug**
Using the `ReadManyItemsAsync` causes thread pool starvation in our application. This is observed by seeing Kestrel emit the log line below a bunch of times (which was also a red flag in [another issue on this repo](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/1852)), health checks going down, and the executing endpoint not finishing.

```
As of "01/22/2023 11:14:20 +00:00", the heartbeat has been running for "00:00:03.4144034" which is longer than "00:00:01". This could be caused by thread pool starvation.
```

**To Reproduce**
I don't have a reproduction scenario at hand, or in a public repo (as this happens as part of a business application with closed sources). However, the full snippet of the code in action is the code below. If more context is needed, and access to our private source is helpful, then I can provide that.

Some numbers:
 * The resulting `budgetIdToEndUserId` dictionary has `1734` items. The `ActivePersonalBudgetSummary` type are a "summary" document that only represents the data we need in this context, the full document is much bigger (we're in the process of splitting these large documents up in smaller ones).
 * size of personalBudgetCollection: `2014`
 * size of activePersonalBudgetCollection: `13186`
 * size of the full dump of these 2 collections: 4.9 GB (with `dt.exe`, and indented)

When we change the `ReadManyItemsAsync()` call to "many `ReadManyItemAsync()`" (no parallelism being done), then the endpoint takes ~1 min.

```csharp
    private async Task<Dictionary<Guid, ActivePersonalBudgetSummary>> GetByPredicateAsync(Func<ActivePersonalBudgetSummary, bool> predicate)
    {
        var personalBudgetCollection = _cosmosClient.GetDatabase(_databaseId).GetContainer(Constants.CosmosDbContainerConstants.PersonalBudgetCollection.ContainerId);
        var activePersonalBudgetCollection = _cosmosClient.GetDatabase(_databaseId).GetContainer(Constants.CosmosDbContainerConstants.ActivePersonalBudgetCollection.ContainerId);

        var personalBudgetIterator = personalBudgetCollection.GetItemQueryIterator<EndUserActivePersonalBudget>(
            "SELECT c.EndUserId, c.ActivePersonalBudgetId FROM c WHERE c.IsSelected");

        var selectedBudgetIds = await personalBudgetIterator.ToListAsync(_monitoringEvents);

        var budgetIdToEndUserId = selectedBudgetIds
            .Where(x => x.ActivePersonalBudgetId is not null)
            .ToDictionary(
                x => x.ActivePersonalBudgetId!,
                x => x.EndUserId);

        var results = new Dictionary<Guid, ActivePersonalBudgetSummary>();
        
        var response = await activePersonalBudgetCollection.ReadManyItemsAsync<ActivePersonalBudgetSummaryDocument>(
            budgetIdToEndUserId.Keys.Select(id => (id, new PartitionKey(id))).ToImmutableList());
        _monitoringEvents?.CosmosDbDiagnosticsReceived(response.Diagnostics);

        foreach (var item in response)
        {
            if (budgetIdToEndUserId.TryGetValue(item.Id, out var endUserId))
            {
                var summary = ToDomain(item);
                if (predicate(summary))
                    results[endUserId] = summary;
            }
        }

        return results;
    }
```

**Expected behavior**
No thread pool starvation happens, and the endpoint returns successfully.

**Actual behavior**
Connecting to the instance using `az webapp create-remote-connection ...`, and running `dotnet counters` as suggested in the [how to detect thread pool starvation ](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/debug-threadpool-starvation#detecting-threadpool-starvation) validates the idea that this is caused by thread pool starvation. The app started out with a thread pool count of `6`, and the highest that I saw was going all the way up to `15` (see screenshots below). Full video can be found [here](https://drive.google.com/file/d/1SltBm0hj28o2E9zGF3kOJk2--EeYsayK/view?usp=share_link).

**At start of the request**
![image-20230123-160440](https://user-images.githubusercontent.com/460904/214297981-a27830ba-3a18-4e4d-a275-2f71eb2d7f68.png)

**After 2:50m**
![image-20230123-160517](https://user-images.githubusercontent.com/460904/214298082-72cfc963-86ff-41d9-9f39-f2bab2604bc0.png)

**Environment summary**
SDK Version: `3.31.2`
OS Version (e.g. Windows, Linux, MacOSX):
 * Azure App Service linux
 * .NET 7 (`7.0.102`)
 * CosmosDB provisioning model: serverless

**Additional context**
We know that the hardware that this app is running on is not production level hardware, from the perspective of App Service. However, this hasn't been an issue so far, so we haven't had any need to scale up to a more production level provisioning. If this bug is related to the restricted environment, then I can fully understand that, and closing down the issue from that angle.

**Comment by ealsur** [comment]
Can you confirm that `GetByPredicateAsync` is being also called through `await` and there is no code that is doing `.Result` or `.GetAwaiter().GetResult()`?

**Comment by rikvandersanden** [comment]
Yes, `GetByPredicateAsync` is always properly awaited

**Comment by tiesmaster** [comment]
Yes, I can confirm. Our call stack is very thin in this context:

```
ActivePersonalBudgetSummariesController.GetByBudgetGroupsAsync(...)
ActivePersonalBudgetSummaryRepository.GetByBudgetGroupsAsync(...)
ActivePersonalBudgetSummaryRepository.GetByPredicateAsync(...)
```

**Comment by tiesmaster** [comment]
Above @rikvandersanden is also working at Buddy Payment, and works primarily on this application. He also pointed out that I wrote:

> When we change the ReadManyItemsAsync() call to "many ReadManyItemAsync()" (no parallelism being done), then the endpoint takes ~1 min.

The second `ReadManyItemAsync()` should have been `ReadItemAsync()`, but that was probably clear from the context.

**Comment by ealsur** [comment]
@NaluTripician  Could you take a look? How does ReadMany work in terms of parallelization?

**Comment by ealsur** [comment]
@tiesmaster Are we certain those threads are actually executing anything inside ReadMany? Is there a dump that can be looked at?

**Comment by tiesmaster** [comment]
@ealsur No, I'm not certain about that, and I can definitely try to make a stack dump of this. However, I don't know exactly on how to do this. I can make a remote connection again with the `az webapp create-remote-connection ...` approach, and use one of the CLI tools. Though, `dotnet-stack` is not available, which looks the most applicable tool for this. `dotnet-trace` and `dotnet-dump` are available there, would this be possible with any of those tools, or is it more convenient to use `dotnet-stack` for this (and in that case, I'll try to get a copy of that on App Service)? Please advise.

**Comment by ealsur** [comment]
Would this be running on Azure App Service? Doesn't it have Diagnostics that might show some information? https://learn.microsoft.com/en-us/azure/app-service/overview-diagnostics

Basically the goal is to identify what are the threads doing to root cause which code is the culprit. @NaluTripician let's check on our end what ReadMany does in terms of parallelization (does it create multiple Tasks? Are they concurrent?). I don't see us using Threads directly in the code.

**Comment by tiesmaster** [comment]
@ealsur Alright, I'll make sure to get to stack trace one way or another. I'm quite busy the next couple of days, so don't expect results before Thursday or Friday

**Comment by NaluTripician** [comment]
Hello, here is some information on the `ReadMany` API that will hopefully provide some insight into your problem. 

First I attempted to recreate your problem on my end by performing a similar query on my end and looking at the threads used for queries of different size. Here are the results from that. 

ReadMany Query Size | Time | Max Thread Count 
-----| ----- | ----- | 
500 | 00:00:03.82 | 7
1000 | 00:00:04.21 | 9
1500 | 00:00:03.61  | 9
2000 | 00:00:04.12 | 10
2001 | 00:00:03.99 | 13
2001 ReadItem Calls | 00:03:11.68 | 10


From these results you can see the `ReadMany` API calls are not causing the max thread count to significantly go up as the number of items read increases and it is executing much faster than multiple `ReadItem` calls.

Now if we look at how the `ReadMany` API works in the SDK in terms of parallelization we can see a few thing. First is that here the SDK manages the [maximum number of requests that can be granted concurrently](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/2608d4af7c45ffb417b72c11d711bca2c570a739/Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs#L26) which is `10*ProcessorCount` (the test I preformed was on a machine that has 12 cores). 10 concurrent request per core is not that high and should not generate thread exhaustion as both the number of concurrent requests and the number of threads is dependent on the number of cores a machine has. 

Next is how are the queries parallelized.  Currently the SDK splits `ReadMany` calls into different queries in two ways. First the SDK will group the items by physical partition. Then within each physical partition it will further split the query into separate queries with no more than [1000 items queried](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/2608d4af7c45ffb417b72c11d711bca2c570a739/Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs#L27).  Splitting by physical partition will make sure that each query is not a cross partition query which will make thread pool starvation even less likely. To see how the queries are split I would recommend looking at [this method](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/2608d4af7c45ffb417b72c11d711bca2c570a739/Microsoft.Azure.Cosmos/src/ReadManyQueryHelper.cs#L81) in ReadManyQueryHelper, which generates the queries and concurrent Tasks that will be executed in parallel.

Please let me know if you have and additional questions about how `ReadMany` works and when you have the stack trace available so I can help find if `ReadMany` is the true culprit behind the thread pool starvation.

**Comment by ealsur** [comment]
@NaluTripician Could you take a look at the logs and see what we can find?

@tiesmaster One small comment on the code snippet, while they achieve the same thing, it is not the same degree of parallelism, the ReadItem alternative is sequential. If you make it truly concurrent, for example:

```
var tasks= new List<Task<ActivePersonalBudgetSummaryDocument>>();
foreach (var (id, partitionKey) in ids)
{
            tasks.Add(this.ReadAndLog((id, partitionKey));
}

await Task.WhenAll(tasks);

// Probably there is a more effective way to return the list, but:
return tasks.Select(t => t.Result);

private async Task<ActivePersonalBudgetSummaryDocument> ReadAndLog()
{
    var result = await activePersonalBudgetCollection.ReadItemAsync<ActivePersonalBudgetSummaryDocument>(id, partitionKey);
    _monitoringEvents?.CosmosDbDiagnosticsReceived(response.Diagnostics);
    return result.Resource;
}

```

What is the result in terms of thread usage? And is it really reaching the machine's max thread count?

**Comment by NaluTripician** [comment]
@tiesmaster I just got a chance to look at the stack traces you sent over. After looking at the baseline and the two others I don't see how this points to thread starvation, but I could be wrong.  In the baseline it looks like one of the threads, `Thread (0x59)` is running code from the Cosmos SDK.  Looking at the other two stack traces it appears that only three threads are running code from the SDK. Let's take a look at what each thread is doing. 

From `stacks1.log` take a look at `Thread (0x59)`:
```
Thread (0x59):
  [Native Frames]
  System.Private.CoreLib!System.Threading.Thread.Sleep(int32)
  Microsoft.Azure.Cosmos.Direct!Microsoft.Azure.Documents.Rntbd.SystemUsageMonitor.RefreshLoopAsync(value class System.Threading.CancellationToken)
  Microsoft.Azure.Cosmos.Direct!Microsoft.Azure.Documents.Rntbd.SystemUsageMonitor.<Start>b__20_0()
  System.Private.CoreLib!System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(class System.Threading.Thread,class System.Threading.ExecutionContext,class System.Threading.ContextCallback,class System.Object)
  System.Private.CoreLib!System.Threading.Tasks.Task.ExecuteWithThreadLocal(class System.Threading.Tasks.Task&,class System.Threading.Thread)
  System.Private.CoreLib!System.Threading.ThreadPoolWorkQueue.Dispatch()
  System.Private.CoreLib!System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
```
Here there are two SDK related things, these are the background CPU monitor, it does not do anything other than snapshot CPU, cannot cause starvation.

Now looking at the other two threads here `Thread (0x1DE)` and `Thread (0x1E4)`. In these two threads are the ones running the query the `ReadMany` call carries out. Both of these queries are not cross partitioned and because there are only two threads running them with many other threads available, it is unlikely that it is Thread Pool starvation.  This is also the case for `Thread (0x208)` in `stacks2.log`. 

The third thread in `stacks2.log`, `Thread (0x220)`, similar to `Thread (0x59)` should not be causing thread pool starvation. From both of the logs it looks like only a few threads are actually running code from the SDK, with many other threads open which is why I am in the opinion that the problem is not thread pool starvation.  

Please let me know if you have any other questions about this. I would be more than happy to answer them and help get to the bottom of this issue.

**Comment by tiesmaster** [comment]
@NaluTripician Thank you for the write-up, and I totally agree with the analysis. This does not look like thread pool starvation, after all. It's interesting that Kestrel triggers that log message as a result of its internal heartbeat, but then that most be result of something else happening in the system.

@ealsur Again, we've solved the problem differently, and moved on, and I don't think there is anything to do on this issue at this stage, so I'll close the issue.. Thank you, and your team for your time, and effort on this.

**Comment by yar-shukan** [comment]
@tiesmaster , can you please share how did you solve the problem: are you still calling ReadManyItemsAsync or you moved away from it to Query/parallel ReadItemAsync?

**Comment by tiesmaster** [comment]
> @tiesmaster , can you please share how did you solve the problem: are you still calling ReadManyItemsAsync or you moved away from it to Query/parallel ReadItemAsync?

@yar-shukan We're not calling anything anymore, as this code was part of a startup, but unfortunately that went out of business. We eventually settled for just calling `ReadItemAsync` in a loop, so we didn't even do this in parallel. We tested running this via PLinq, or `Parallel.For`, but that didn't really move the needle, so we just kept a simple `foreach` loop.

I hope that helps. Let me know if you have any further questions. Good luck!
